src/main/java/org/opensearch/searchrelevance/executors/ExperimentRunningManager.java

```diff
@@ -0,0 +1,601 @@
+    }
+
+    @VisibleForTesting
+    void fetchSearchConfigurationsAsync(
+        String experimentId,
+        PutExperimentRequest request,
+        List<String> queryTextWithReferences,
+        ExperimentCancellationToken cancellationToken,
+        CountDownLatch actuallyFinished
+    ) {
+        Map<String, SearchConfigurationDetails> searchConfigurations = new HashMap<>();
+        AtomicBoolean hasFailure = new AtomicBoolean(false);
+        List<CompletableFuture<Entry<String, Object>>> configFutures = new ArrayList<>();
+
+        for (String configId : request.getSearchConfigurationList()) {
+            CompletableFuture<Entry<String, Object>> singleSearchConfigurationFuture = fetchSingleSearchConfigurationAsync(
+                experimentId,
+                request,
+                queryTextWithReferences,
+                hasFailure,
+                configId,
+                cancellationToken,
+                actuallyFinished
+            );
+            // We will have to set futures to be cancelled in case of timeout.
+            if (request.getScheduledExperimentResultId() != null && checkIfCancelled(cancellationToken) == false) {
+                runningFutures.get(request.getScheduledExperimentResultId()).add(singleSearchConfigurationFuture);
+            }
+            configFutures.add(singleSearchConfigurationFuture);
+        }
+
+        for (CompletableFuture<Entry<String, Object>> configFuture : configFutures) {
+            Entry<String, Object> configEntry;
+            try {
+                // The config future will be waited on, but there is a chance the future might be null.
+                configEntry = configFuture.get();
+            } catch (InterruptedException e) {
+                handleAsyncFailure(experimentId, request, "Failed to fetch SearchConfiguration", e, actuallyFinished);
+                return;
+            } catch (ExecutionException e) {
+                handleAsyncFailure(experimentId, request, "Failed to fetch SearchConfiguration", e, actuallyFinished);
+                return;
+            }
+            if (configEntry == null) {
+                throw new IllegalStateException("Experiment has been timed out while search configuration fetching or there was a failure");
+            }
+            searchConfigurations.put(configEntry.getKey(), (SearchConfigurationDetails) configEntry.getValue());
+        }
+
+        if (queryTextWithReferences == null || searchConfigurations == null) {
+            throw new IllegalStateException("Missing required data for metrics calculation");
+        }
+
+        List<Map<String, Object>> finalResults = Collections.synchronizedList(new ArrayList<>());
+        AtomicInteger pendingQueries = new AtomicInteger(queryTextWithReferences.size());
+
+        executeExperimentEvaluation(
+            experimentId,
+            request,
+            searchConfigurations,
+            queryTextWithReferences,
+            finalResults,
```

Comment:
looks like we're simply accumulating results into this List of maps `finalResults`. This could consume significant memory. Let's at least signal to the logs in case we reached certain high threshold, probably Runtime class can give some helpful info: `Runtime.getRuntime().totalMemory()` or `Runtime.getRuntime().freeMemory()`